File: old_backend/campaignResearch.js
import ModelClient, { isUnexpected } from '@azure-rest/ai-inference';
import { AzureKeyCredential } from '@azure/core-auth';
import { getCategoryTaxonomyForPrompt, resolveCategoryFromResearch } from './categories.js';

const modelName = 'openai/gpt-4o-mini';
let client = null;

function getClient() {
  if (!client) {
    const token = process.env.GITHUB_TOKEN;
    if (!token) return null;
    client = ModelClient('https://models.github.ai/inference', new AzureKeyCredential(token));
  }
  return client;
}

/**
 * AI-powered campaign research — barcha forma maydonlarini to‘ldiradi.
 */
export async function researchCampaign({ brand_url, name, category, brief }) {
  const ai = getClient();
  if (!ai) {
    return { error: 'GITHUB_TOKEN sozlanmagan — AI tadqiqot mavjud emas' };
  }

  let pageHint = '';
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10000);
    const res = await fetch(brand_url, {
      signal: controller.signal,
      headers: { 'User-Agent': 'SynapticBot/1.0 (+https://synaptic.uz)' },
      redirect: 'follow',
    });
    clearTimeout(timeout);
    if (res.ok) {
      const html = await res.text();
      const title = html.match(/<title[^>]*>([^<]+)<\/title>/i)?.[1]?.trim();
      const desc = html.match(/<meta[^>]+name=["']description["'][^>]+content=["']([^"']+)["']/i)?.[1]?.trim();
      const snippet = html
        .replace(/<script[\s\S]*?<\/script>/gi, '')
        .replace(/<[^>]+>/g, ' ')
        .replace(/\s+/g, ' ')
        .trim()
        .slice(0, 3500);
      pageHint = [title && `Title: ${title}`, desc && `Description: ${desc}`, snippet && `Content: ${snippet}`]
        .filter(Boolean)
        .join('\n');
    }
  } catch {
    pageHint = '(Sayt kontenti avtomatik o‘qilmadi — URL va qisqa izoh asosida tahlil qilinadi)';
  }

  const taxonomyJson = JSON.stringify(getCategoryTaxonomyForPrompt(), null, 0);

  const systemPrompt = `You are a digital marketing strategist for Uzbekistan (O'zbekistan).
Analyze the brand website and output ONLY valid JSON (no markdown fences) with ALL keys:

name (string, brand/campaign title in Uzbek or Russian),
tagline (string, short ad slogan),
description (string, 2-4 sentences for advertisers),
keywords (array of 10-18 phrases users type in AI chat when looking for this product — Uzbek/Russian mix),
niche_keywords (array of 4-8 long-tail phrases),
link_text (string, CTA button max 4 words, e.g. "Batafsil ko'rish"),
suggested_category (string, MUST be one category id from taxonomy below),
suggested_subcategory (string, MUST be valid subcategory id for that category),
custom_category (string, only if suggested_category is "other", else ""),
tone (one of: informative, promotional, comparative, deal-focused),
cpc_rate_suggestion (integer UZS per click, realistic 800-12000),
budget_suggestion (integer UZS total campaign budget, realistic 500000-50000000),
offerings (array of 3-12 items the business sells — each object):
  { name, type ("product" or "service"), description (1 sentence), url (product/deep link or empty),
    price_hint (e.g. "1.2 mln so'm" or ""), keywords (array 3-6 search phrases for that item) },
research_summary (2-3 sentences in Uzbek).

TAXONOMY (use exact ids):
${taxonomyJson}

Rules:
- Pick the closest category/subcategory from taxonomy
- keywords must match how Uzbeks ask AI assistants (prices, where to buy, delivery, installment)
- budget_suggestion should match business size implied by the site`;

  const userContent = `Brand URL: ${brand_url}
${name ? `Known name: ${name}` : ''}
${category ? `Category hint: ${category}` : ''}
${brief ? `Owner brief: ${brief}` : ''}

Page data:
${pageHint || 'No page scrape'}`;

  try {
    const response = await ai.path('/chat/completions').post({
      body: {
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userContent },
        ],
        model: modelName,
        temperature: 0.35,
        max_tokens: 2400,
      },
    });

    if (isUnexpected(response)) {
      throw new Error(response.body?.error?.message || 'AI javob bermadi');
    }

    const raw = response.body.choices[0]?.message?.content?.trim() || '';
    const jsonMatch = raw.match(/\{[\s\S]*\}/);
    if (!jsonMatch) throw new Error('AI JSON qaytarmadi');

    const parsed = JSON.parse(jsonMatch[0]);
    const { category: catId, subcategory: subId } = resolveCategoryFromResearch(
      parsed.suggested_category,
      parsed.suggested_subcategory
    );

    return {
      name: String(parsed.name || name || '').slice(0, 120),
      tagline: String(parsed.tagline || '').slice(0, 200),
      description: String(parsed.description || '').slice(0, 1000),
      keywords: Array.isArray(parsed.keywords) ? parsed.keywords.map(String).slice(0, 20) : [],
      niche_keywords: Array.isArray(parsed.niche_keywords)
        ? parsed.niche_keywords.map(String).slice(0, 10)
        : [],
      link_text: String(parsed.link_text || 'Batafsil').slice(0, 40),
      category: catId,
      subcategory: subId,
      custom_category: catId === 'other' ? String(parsed.custom_category || '').slice(0, 80) : '',
      tone: ['informative', 'promotional', 'comparative', 'deal-focused'].includes(parsed.tone)
        ? parsed.tone
        : 'informative',
      cpc_rate_suggestion: Math.max(500, Number(parsed.cpc_rate_suggestion) || 2500),
      budget_suggestion: Math.max(100000, Number(parsed.budget_suggestion) || 2000000),
      research_summary: String(parsed.research_summary || ''),
      offerings: Array.isArray(parsed.offerings)
        ? parsed.offerings
            .map((o) => ({
              name: String(o.name || '').slice(0, 120),
              type: o.type === 'service' ? 'service' : 'product',
              description: String(o.description || '').slice(0, 500),
              url: String(o.url || brand_url || '').slice(0, 500),
              price_hint: String(o.price_hint || '').slice(0, 80),
              keywords: Array.isArray(o.keywords)
                ? o.keywords.map(String).slice(0, 12)
                : [],
            }))
            .filter((o) => o.name)
            .slice(0, 12)
        : [],
    };
  } catch (err) {
    console.error('Campaign research failed:', err.message);
    return { error: err.message || 'AI tadqiqot muvaffaqiyatsiz' };
  }
}

File: backend/app/services/campaign_research.py
import json
import logging
import re

import httpx
from openai import AsyncOpenAI

from app.config import settings
from app.services.categories import get_category_taxonomy_for_prompt, resolve_category_from_research

logger = logging.getLogger(__name__)
MODEL = "openai/gpt-4o-mini"


def _client() -> AsyncOpenAI | None:
    if not settings.github_token:
        return None
    return AsyncOpenAI(base_url="https://models.github.ai/inference", api_key=settings.github_token)


async def research_campaign(*, brand_url: str, name: str = "", category: str = "", brief: str = "") -> dict:
    ai = _client()
    if not ai:
        return {"error": "GITHUB_TOKEN sozlanmagan — AI tadqiqot mavjud emas"}

    page_hint = ""
    try:
        async with httpx.AsyncClient(timeout=10, follow_redirects=True) as client:
            res = await client.get(brand_url, headers={"User-Agent": "SynapticBot/1.0 (+https://synaptic.uz)"})
            if res.status_code == 200:
                html = res.text
                title = re.search(r"<title[^>]*>([^<]+)</title>", html, re.I)
                desc = re.search(r'<meta[^>]+name=["\']description["\'][^>]+content=["\']([^"\']+)["\']', html, re.I)
                snippet = re.sub(r"<script[\s\S]*?</script>", "", html, flags=re.I)
                snippet = re.sub(r"<[^>]+>", " ", snippet)
                snippet = re.sub(r"\s+", " ", snippet).strip()[:3500]
                parts = []
                if title:
                    parts.append(f"Title: {title.group(1).strip()}")
                if desc:
                    parts.append(f"Description: {desc.group(1).strip()}")
                if snippet:
                    parts.append(f"Content: {snippet}")
                page_hint = "\n".join(parts)
    except Exception:
        page_hint = "(Sayt kontenti avtomatik o‘qilmadi — URL va qisqa izoh asosida tahlil qilinadi)"

    taxonomy = json.dumps(get_category_taxonomy_for_prompt(), ensure_ascii=False)
    system = f"""You are a digital marketing strategist for Uzbekistan.
Output ONLY valid JSON with keys: name, tagline, description, keywords, niche_keywords, link_text,
suggested_category, suggested_subcategory, custom_category, tone, cpc_rate_suggestion, budget_suggestion,
offerings (array of {{name,type,description,url,price_hint,keywords}}), research_summary.
TAXONOMY: {taxonomy}"""

    user = f"Brand URL: {brand_url}\nKnown name: {name}\nCategory hint: {category}\nBrief: {brief}\nPage:\n{page_hint}"

    try:
        resp = await ai.chat.completions.create(
            model=MODEL,
            messages=[{"role": "system", "content": system}, {"role": "user", "content": user}],
            temperature=0.35,
            max_tokens=2400,
        )
        raw = resp.choices[0].message.content or ""
        m = re.search(r"\{[\s\S]*\}", raw)
        if not m:
            raise ValueError("AI JSON qaytarmadi")
        parsed = json.loads(m.group(0))
        cat_id, sub_id = resolve_category_from_research(parsed.get("suggested_category"), parsed.get("suggested_subcategory"))
        return {
            "name": str(parsed.get("name") or name or "")[:120],
            "tagline": str(parsed.get("tagline") or "")[:200],
            "description": str(parsed.get("description") or "")[:2000],
            "keywords": (parsed.get("keywords") or [])[:40],
            "niche_keywords": (parsed.get("niche_keywords") or [])[:20],
            "link_text": str(parsed.get("link_text") or "Batafsil")[:40],
            "category": cat_id,
            "subcategory": sub_id,
            "custom_category": str(parsed.get("custom_category") or "")[:80],
            "tone": parsed.get("tone") or "informative",
            "cpc_rate_suggestion": int(parsed.get("cpc_rate_suggestion") or 2000),
            "budget_suggestion": int(parsed.get("budget_suggestion") or 5_000_000),
            "offerings": parsed.get("offerings") or [],
            "research_summary": str(parsed.get("research_summary") or ""),
        }
    except Exception as err:
        logger.exception("Campaign research failed")
        return {"error": str(err)}
